### DISCUSSION TITLE
using different form of cost function instead of fidelity

### DISCUSSION THREAD

**Discussion by Leeseok-628** [qucontrol/krotov#99] [Q&A] [answered]
Hi, 

I have been wondering if I can use krotov to optimize different forms of cost function, other than fidelity. 
For instance, would it be possible to optimize trace(rho_target - rho_ansatz) where rho_target is target density matrix and rho_ansatz is the time-evolved state under control parameters. If so, how?

Additionally, suppose we have some random operator A. Then is it possible to minimize the trace distance between the two density operators mutilplied by this operator A: i.e. trace(A(rho_target - rho_ansatz)). Does this also feasible via krotov?

**Comment by christianekoch** [comment]
Of course!
Your request corresponds to "Optimization of a Dissipative State-to-State Transfer in a Lambda System" listed as one of the examples in the package documentation. 
In case of a mixed target state, please see  https://doi.org/10.1002/qute.201800110 for potential pitfalls in defining your target functional.

**Comment by christianekoch** [comment]
The answer to your second question is also, yes. This would simply be the optimization of an expectation value difference which is one of the standard optimization tasks.

**Comment by Leeseok-628** [comment]
Thank you so much for your response! I read through the example, it helped me a lot. 

How can I customize (or build) my own cost function that goes into `krotov.optimize_pulses` function? Regarding my second question earlier, as you mentioned above, the cost function is tr(D(rho_tar - rho_trial)) = <psi_trial|D|psi_trial> - <psi_target|D|psi_target>, for given D and psi_target. But I don't see how to customize/code the cost function in this form of summation of two expectation values. The objectives that goes into `krotov.optimize_pulses` requires one to set initial and final states within some Hamiltonian, but I'm not quite sure how to build objectives when I need to calculate the expectation value of given operator D as my cost function. I'm confused on this because, I looked into the details of `krotov functional` (https://qucontrol.github.io/krotov/v1.0.0/API/krotov.functionals.html#module-krotov.functionals), and it seems like all functionals don't deal with expectation values nor linear combinations of two distinct terms! 

I'd really appreciate it if you can help me out with this :)

**Comment by goerz** [comment, marked as answer]
From a technical perspective, the only thing required for a new functional is to define a function that evaluates `∂J/∂⟨ϕₖ|` and that you can pass as `chi_constructor` to the `optimize_pulses` routine, as outlined in the documentation of [`krotov.functionals`](https://qucontrol.github.io/krotov/master/API/krotov.functionals.html#).

I would also point out that [`J_T_hs`](https://qucontrol.github.io/krotov/master/API/krotov.functionals.html#krotov.functionals.J_T_hs) / [`chis_hs`](https://qucontrol.github.io/krotov/master/API/krotov.functionals.html#krotov.functionals.chis_hs) is already defined and should cover your first case. For the second case of the expectation value, you'll have to write your own function, but it should be fairly straightforward to calculate the χ-states exactly analogous to the calculation shown in the [`chis_hs` docstring](https://qucontrol.github.io/krotov/master/API/krotov.functionals.html#krotov.functionals.chis_hs).

If you find the matrix calculus notation unsettling, `∂J/∂⟨ϕₖ|` is just shorthand for a [Wirtinger derivative](https://qucontrol.github.io/krotov/master/07_krotovs_method.html#equation-wirtinger-deriv), but calculating it component-wise is extremely cumbersome. You should be fine just writing out your functional in terms of inner products in Liouville space (`Tr[A⁺B] ≡ ⟨A|B⟩`) and taking the derivative with respect to the co-state `⟨A|` treated as an independent variable, as we've done for all the existing functionals.

**Reply by Leeseok-628** [reply]
Thank you so much for your response!

To make it crystal clear, if I want to minimize J = -⟨ϕ(T)|O|ϕ(T)> for given operator O (no target state), then I might have to write a function of |chi_k> = ∂J/∂⟨ϕ_k(T)|, which I believe is just simply -O|ϕ(T)>? 

Then, when defining the `objectives`, I might leave the `target` as None, since my cost function J doens't have target terms. 

Does this sound reasonable to you?

**Reply by goerz** [reply]
Yeah, that sounds good. You might want to put `O` in `target`, otherwise, you'd have to hard-code it.

**Reply by goerz** [reply]
Oh, and it's `χ = - ∂J/∂⟨ϕₖ|` (note that minus sign!), so that would be `O |ϕₖ(T)⟩`, not `-O |ϕₖ(T)⟩`

**Reply by Leeseok-628** [reply]
Great, thanks you :) 
For some reason, it's not getting trained properly, the cost function flucuates very much. 

Can we bound intervals of optimized pulse amplidues (e.g. [-1, 1]), just like qutip optimize pusle's amp_lbound/ubound?

**Reply by goerz** [reply]
Increase your [`λₐ`](https://qucontrol.github.io/krotov/master/API/krotov.optimize.html?highlight=lambda_a).

If you don't get monotonic convergence even for large `λₐ`, there's something wrong with your `chi_constructor`.

As for limiting the amplitude, you'd have to use parametrized pulses, which is work-in-progress at https://github.com/qucontrol/krotov/pull/92. Feel free to play around with that, or try the workaround in https://github.com/qucontrol/krotov/issues/23#issuecomment-850887732.

**Reply by Leeseok-628** [reply]
Thank you! 

In my problem of interest, there's no need to specify parametrizations for the pulse functions. It's enough to make the pulse amplitudes are in [-1, 1] during the optimization process, no matter what shapes of the pulse amplitudes are optimized. Isn't there an easier way to achieve this condition? (since it seems like parametrized pulses is not yet updated to the Krotov doc)

**Reply by goerz** [reply]
The only other options would be to use the `modify_params_after_iter` argument and pass a function that simply truncates the amplitude after each iteration. All bets on monotonic convergence are off if you do that, and the resulting control may not be physical, due to the non-smooth cutoff.

Using parametrization should be quite straightforward, even while it hasn't yet made it into a release:

* Install the [`23-pulse-parametrization`](https://github.com/qucontrol/krotov/tree/23-pulse-parameterization) branch: `pip install git+https://github.com/qucontrol/krotov.git@23-pulse-parameterization`
* [Follow the example notebook illustrating parametrized pulses](https://nbviewer.org/github/qucontrol/krotov/blob/23-pulse-parameterization/docs/notebooks/01_example_simple_state_to_state_parametrization.ipynb)

**Reply by Leeseok-628** [reply]
I see, thanks! So `TanhParametrization` would be the way if I have to bound my eps_max and min (e.g. every pulse amplitudes in [-1, 1] would be just eps_max = 1, eps_min = -1).

Intuitively, I am curious to know the reason behind using hyperbolic tangent as the function to bound the pulse amplitudes. (as shown in https://qucontrol.github.io/krotov/v1.2.1/10_howto.html#how-to-limit-the-amplitude-of-the-controls) Could you give me a brief explanation? (or any related literatures would be very helpful!)

**Reply by goerz** [reply]
Yes, that's right.

As for why the `tanh`: limiting the amplitude to an open interval implies some kind of [sigmoid function](https://en.wikipedia.org/wiki/Sigmoid_function). You could pretty much use any sigmoid function you want, but `tanh` has the nice property that it doesn't introduce any distortions in the update (the leading term in the Taylor series is `x`). It's equivalent to a [Logistic function](https://en.wikipedia.org/wiki/Logistic_function) with `k=2`. You could [play around](https://juliaquantumcontrol.github.io/Krotov.jl/dev/examples/state_to_state_parametrizations/#Symmetric-Bounded-Controls) with other logistic parameters if you want to influence how sharp you want the amplitude cutoff to be, but then you might have to compensate the resulting distortion by scaling λₐ.

As far as I'm aware, the use of the `tanh` parametrization in quantum control goes back to [Müller et al. Quantum Inf. Process. 10, 771 (2011)](http://dx.doi.org/10.1007/s11128-011-0296-0).

**Reply by Leeseok-628** [reply]
If J is a trace distance between the two states, i.e. `J = Tr(|ϕ(T)><ϕ(T)| - |ϕ_targ><ϕ_targ|)`, what would `χ` be?

**Reply by goerz** [reply]
I don't know! You might actually have to look at the component-wise Wirtinger derivative if you can't write your functional as some kind of inner product. Are you sure you don't just want to use `||ρ(T) - ρ_targ||` defined via the Hilbert-Schmidt inner product?

@danielreich Anything to add?

**Reply by Leeseok-628** [reply]
I see, I might give it a shot. 

Also, is there any good way/intuition on how to choose S(t) (https://qucontrol.github.io/krotov/v1.2.1/API/krotov.optimize.html?highlight=update_shape)? I'm going over example notebooks in Krotov, and I don't see any clear reason on choosing such different S(t)s for different examples.

**Reply by goerz** [reply]
No, there's no particular intuition. You can use whatever you want.

**Reply by Leeseok-628** [reply]
Sorry for asking lots of questions.. This might be my last question for a while.

I'm planning to minimize the cost function that has an absolute value form, which is` J = |<ϕ(T)|O|ϕ(T)> - C|`, where C is just a given constant (so ultimately, by minimizing J to 0, I want to make **<ϕ(T)|O|ϕ(T)> as close as C**). 

But I confront two problems:
1) For different constant C, `χ = - ∂J/∂⟨ϕₖ|` would be the same, right? So, how could we define such forms cost functions that have constant terms, by only defining `χ`? (or maybe it'd be great if Krotov is possible to break the minimization whenever J = C or `isclose(J,C)=True`)
2) In this case of absolute valued cost function, how could we build `χ`? (J seems like -|O|ϕ(T)>|, but it's definitely not a square root of matrix since O|ϕ(T)> is not a square matrix).

**Reply by goerz** [reply]
First, you always want to use the square of the norm instead of just the norm, to get rid of the square root.. Then `χ` is definitely not going to be independent of `C`: you'll have `(⟨ϕ(T)|O|ϕ(T)⟩ - C)(⟨ϕ(T)|O|ϕ(T)⟩ - C)^†`, which you have to multiply out and take the derivative of w.r.t. `⟨ϕ(T)|`

**Reply by Leeseok-628** [reply]
I see, thanks! So, since O is Hermitian, ∂J/∂⟨ϕ(T)| will be just `2(⟨ϕ(T)|O|ϕ(T)⟩-C) * O|ϕ(T)⟩` !

**Reply by goerz** [reply]
That looks right

**Reply by Leeseok-628** [reply]
Awesome, so I assume the reason of avoiding the square root is that then we can't take derivative w.r.t `⟨ϕ(T)|`; so that setting `J = |<ϕ(T)|O|ϕ(T)> - C|` would be impossible to optimize through Krotov.

**Reply by goerz** [reply]
It's not *impossible* (you can take the derivative of the square root), but it's an unnecessary complication. Minimizing `|x|` and minimizing `|x|²` is mathematically equivalent, and gradients of `|x|²` are always going to be more straightforward
